# 4.8: Grouping Data & Aggregating Variables

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Creating a path 
path = r'C:\Users\fairo\03-2023 Instacart Basket Analysis'

In [3]:
# Importing data
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_update.pkl')) 

In [4]:
# Creating a subset 
df = ords_prods_merge[:1000000]

In [5]:
ords_prods_merge = ords_prods_merge.drop(columns = ['_merge'])

In [6]:
df.shape

(1000000, 18)

In [7]:
df.head(10)

order_id  user_id  order_number  orders_day_of_week  orders_hour_of_day  \
0   2539329        1             1                   2                   8   
1   2398795        1             2                   3                   7   
2    473747        1             3                   3                  12   
3   2254736        1             4                   4                   7   
4    431534        1             5                   4                  15   
5   3367565        1             6                   2                   7   
6    550135        1             7                   1                   9   
7   3108588        1             8                   1                  14   
8   2295261        1             9                   1                  16   
9   2550362        1            10                   4                   8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   
5                    19.0         196                  1          1   both   
6                    20.0         196                  1          1   both   
7                    14.0         196                  2          1   both   
8                     0.0         196                  4          1   both   
9                    30.0         196                  1          1   both   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   
5         Soda        77              7     9.0  Mid-range product   
6         Soda        77              7     9.0  Mid-range product   
7         Soda        77              7     9.0  Mid-range product   
8         Soda        77              7     9.0  Mid-range product   
9         Soda        77              7     9.0  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day  
0  Regularly busy  Regularly busy        Average orders  
1  Regularly busy    Slowest days        Average orders  
2  Regularly busy    Slowest days           Most orders  
3      Least busy    Slowest days        Average orders  
4      Least busy    Slowest days           Most orders  
5  Regularly busy  Regularly busy        Average orders  
6  Regularly busy    Busiest days        Average orders  
7  Regularly busy    Busiest days           Most orders  
8  Regularly busy    Busiest days        Average orders  
9      Least busy    Slowest days        Average orders

In [8]:
# Grouping Data with pandas
df.groupby('product_name')

In [9]:
# Aggregating Data with agg()
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

In [10]:
df.groupby('department_id')['order_number'].mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

In [11]:
# Performing Multiple Aggregations
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

In [12]:
# Aggregating Data with transform()
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [13]:
ords_prods_merge.head(15)

order_id  user_id  order_number  orders_day_of_week  orders_hour_of_day  \
0    2539329        1             1                   2                   8   
1    2398795        1             2                   3                   7   
2     473747        1             3                   3                  12   
3    2254736        1             4                   4                   7   
4     431534        1             5                   4                  15   
5    3367565        1             6                   2                   7   
6     550135        1             7                   1                   9   
7    3108588        1             8                   1                  14   
8    2295261        1             9                   1                  16   
9    2550362        1            10                   4                   8   
10   2968173       15            15                   1                   9   
11   1870022       15            17                   2                  16   
12   1911383       15            18                   2                  11   
13   2715276       15            21                   1                   9   
14    487368       15            22                   1                  10   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   

   product_name  aisle_id  department_id  prices    price_range_loc  \
0          Soda        77              7     9.0  Mid-range product   
1          Soda        77              7     9.0  Mid-range product   
2          Soda        77              7     9.0  Mid-range product   
3          Soda        77              7     9.0  Mid-range product   
4          Soda        77              7     9.0  Mid-range product   
5          Soda        77              7     9.0  Mid-range product   
6          Soda        77              7     9.0  Mid-range product   
7          Soda        77              7     9.0  Mid-range product   
8          Soda        77              7     9.0  Mid-range product   
9          Soda        77              7     9.0  Mid-range product   
10         Soda        77              7     9.0  Mid-range product   
11         Soda        77              7     9.0  Mid-range product   
12         Soda        77              7     9.0  Mid-range product   
13         Soda        77              7     9.0  Mid-range product   
14         Soda        77              7     9.0  Mid-range product   

       busiest_day    busiest_days busiest_period_of_day  max_order  
0   Regularly busy  Regularly busy        Average orders         10  
1   Regularly busy    Slowest days        Average orders         10  
2   Regularly busy    Slowest days           Most orders         10  
3       Least busy    Slowest days        Average orders         10  
4       Least busy    Slowest days           Most orders         10  
5   Regularly busy  Regularly

In [14]:
# Creating loyalty Flag
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [15]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [16]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [17]:
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [18]:
# Checking 
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head()

user_id  loyalty_flag  order_number
0        1  New customer             1
1        1  New customer             2
2        1  New customer             3
3        1  New customer             4
4        1  New customer             5

# Task

In [19]:
2.# Finding the aggregated mean for the entire dataframe
ords_prods_merge.groupby('department_id')['order_number'].mean()

department_id
1     15.457838
2     17.277920
3     17.170395
4     17.811403
5     15.215751
6     16.439806
7     17.225802
8     15.340650
9     15.895474
10    20.197148
11    16.170638
12    15.887671
13    16.583536
14    16.773669
15    16.165037
16    17.665606
17    15.694469
18    19.310397
19    17.177343
20    16.473447
21    22.902379
Name: order_number, dtype: float64

The results for the entire dataframe are different from those of the subset because it contains all the departments contrary to the subset, so the mean of the entire dataframe is more signifiant than the subset one. 

In [20]:
5.# Checking the basic statistics of the products prices for each loyalty category 
print('The mean, min and max of prices grouped by the loyalty_flag column in ords_prods_merge:')
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

The mean, min and max of prices grouped by the loyalty_flag column in ords_prods_merge:


prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

New customers tend to go for the highest prices, followed by regular customers and finally loyal customers, who buy the least expensive products. However, the maximum price is 99,999, it maybe going to skewed the distribution.

In [21]:
6.# Create a spending flag for each user based on the average price across all their orders 
# creating “spending_flag” column using transform()
ords_prods_merge['average_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [22]:
# Rounding averages 
ords_prods_merge['average_price'] = ords_prods_merge['average_price'].round(2)

In [23]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  orders_hour_of_day  \
0   2539329        1             1                   2                   8   
1   2398795        1             2                   3                   7   
2    473747        1             3                   3                  12   
3   2254736        1             4                   4                   7   
4    431534        1             5                   4                  15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

      busiest_day    busiest_days busiest_period_of_day  max_order  \
0  Regularly busy  Regularly busy        Average orders         10   
1  Regularly busy    Slowest days        Average orders         10   
2  Regularly busy    Slowest days           Most orders         10   
3      Least busy    Slowest days        Average orders         10   
4      Least busy    Slowest days           Most orders         10   

   loyalty_flag  average_price  
0  New customer           6.37  
1  New customer           6.37  
2  New customer           6.37  
3  New customer           6.37  
4  New customer           6.37

In [24]:
# Creating a 'spending_flag' 
ords_prods_merge.loc[ords_prods_merge['average_price'] < 10, 'spending_flag'] = 'Low spender'

In [25]:
ords_prods_merge.loc[ords_prods_merge['average_price'] >= 10, 'spending_flag'] = 'High spender'

In [26]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  orders_hour_of_day  \
0   2539329        1             1                   2                   8   
1   2398795        1             2                   3                   7   
2    473747        1             3                   3                  12   
3   2254736        1             4                   4                   7   
4    431534        1             5                   4                  15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  department_id  prices    price_range_loc     busiest_day  \
0         Soda  ...              7     9.0  Mid-range product  Regularly busy   
1         Soda  ...              7     9.0  Mid-range product  Regularly busy   
2         Soda  ...              7     9.0  Mid-range product  Regularly busy   
3         Soda  ...              7     9.0  Mid-range product      Least busy   
4         Soda  ...              7     9.0  Mid-range product      Least busy   

     busiest_days busiest_period_of_day max_order  loyalty_flag average_price  \
0  Regularly busy        Average orders        10  New customer          6.37   
1    Slowest days        Average orders        10  New customer          6.37   
2    Slowest days           Most orders        10  New customer          6.37   
3    Slowest days        Average orders        10  New customer          6.37   
4    Slowest days           Most orders        10  New customer          6.37   

   spending_flag  
0    Low spender  
1    Low spender  
2    Low spender  
3    Low spender  
4    Low spender  

[5 rows x 21 columns]

In [28]:
# Checking counts
ords_prods_merge['spending_flag'].value_counts()

Low spender     31769965
High spender      634894
Name: spending_flag, dtype: int64

In [29]:
7.# Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. 
ords_prods_merge['median_prior_orders'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [30]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  orders_hour_of_day  \
0   2539329        1             1                   2                   8   
1   2398795        1             2                   3                   7   
2    473747        1             3                   3                  12   
3   2254736        1             4                   4                   7   
4    431534        1             5                   4                  15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  prices    price_range_loc     busiest_day  \
0         Soda  ...     9.0  Mid-range product  Regularly busy   
1         Soda  ...     9.0  Mid-range product  Regularly busy   
2         Soda  ...     9.0  Mid-range product  Regularly busy   
3         Soda  ...     9.0  Mid-range product      Least busy   
4         Soda  ...     9.0  Mid-range product      Least busy   

     busiest_days busiest_period_of_day max_order  loyalty_flag  \
0  Regularly busy        Average orders        10  New customer   
1    Slowest days        Average orders        10  New customer   
2    Slowest days           Most orders        10  New customer   
3    Slowest days        Average orders        10  New customer   
4    Slowest days           Most orders        10  New customer   

   average_price spending_flag  median_prior_orders  
0           6.37   Low spender                 20.5  
1           6.37   Low spender                 20.5  
2           6.37   Low spender                 20.5  
3           6.37   Low spender                 20.5  
4           6.37   Low spender                 20.5  

[5 rows x 22 columns]

In [31]:
ords_prods_merge.loc[ords_prods_merge['median_prior_orders'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [32]:
ords_prods_merge.loc[(ords_prods_merge['median_prior_orders'] > 10) & (ords_prods_merge['median_prior_orders'] <= 20), 'order_frequency_flag'] = 'Regular customer'

In [33]:
ords_prods_merge.loc[ords_prods_merge['median_prior_orders'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [34]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  orders_hour_of_day  \
0   2539329        1             1                   2                   8   
1   2398795        1             2                   3                   7   
2    473747        1             3                   3                  12   
3   2254736        1             4                   4                   7   
4    431534        1             5                   4                  15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...    price_range_loc     busiest_day    busiest_days  \
0         Soda  ...  Mid-range product  Regularly busy  Regularly busy   
1         Soda  ...  Mid-range product  Regularly busy    Slowest days   
2         Soda  ...  Mid-range product  Regularly busy    Slowest days   
3         Soda  ...  Mid-range product      Least busy    Slowest days   
4         Soda  ...  Mid-range product      Least busy    Slowest days   

  busiest_period_of_day max_order  loyalty_flag average_price  spending_flag  \
0        Average orders        10  New customer          6.37    Low spender   
1        Average orders        10  New customer          6.37    Low spender   
2           Most orders        10  New customer          6.37    Low spender   
3        Average orders        10  New customer          6.37    Low spender   
4           Most orders        10  New customer          6.37    Low spender   

  median_prior_orders   order_frequency_flag  
0                20.5  Non-frequent customer  
1                20.5  Non-frequent customer  
2                20.5  Non-frequent customer  
3                20.5  Non-frequent customer  
4                20.5  Non-frequent customer  

[5 rows x 23 columns]

In [35]:
# Checking counts
ords_prods_merge['spending_flag'].value_counts()

Low spender     31769965
High spender      634894
Name: spending_flag, dtype: int64

In [36]:
ords_prods_merge.shape

(32404859, 23)

In [37]:
9.# Exporting the dataframe
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_updated_1.pkl'))